### PR TITLE
Display Results calculated by BOOST

### PR DESCRIPTION
**Pull request by kdav108**
## Description

This PR adds a UI component `BoostResult` to display the output from BOOST when a power plan is generated. This includes the maximum achievable speed and the zone distance and recommend power in that zone to achieve the max speed.

## Screenshots

When no power plan generated:
Screen Shot 2021-07-17 at 9 10 47 pm

When power plan generated:
Screen Shot 2021-07-17 at 9 11 17 pm

## Steps to Test

1) Run both client and server
1.5) You may need to run
```
mosquitto_pub -t "boost/configs" -h localhost -m '{"rider": [{"displayName": "Default rider", "fileName": "default_rider.json"}], "track": [{"displayName": "Ford", "fileName": "ford.json"}, {"displayName": "Holden", "fileName": "holden.json"}, {"displayName": "Battle Mountain", "fileName": "bm.json"}], "bike": [{"displayName": "Wombat", "fileName": "v2_old_camera.json"}, {"displayName": "Blacksmith", "fileName": "v1.5.json"}, {"displayName": "Wombat", "fileName": "v2.json"}, {"displayName": "Priscilla", "fileName": "v3.json"}], "powerPlan": [{"displayName": "test", "fileName": "test.json"}, {"displayName": "default", "fileName": "default.json"}]}'
```
to populate the configs on the Dashboard to use the `generate` button.

2) Use another terminal and publish to `boost/generate_complete` topic. Shortcut:
```
mosquitto_pub -t 'boost/generate_complete' -h localhost -m "{\"fileName\":\"example-File-Name2\",\"maxSpeed\":25,\"zones\":[{\"power\":20,\"distance\":25},{\"power\":30,\"distance\":40}]}"
```
3) Watch the results displayed to the UI

### CODE REVIEW

**Comment by haotongwang** [comment]
Good work on the PR!
The main thing I can see is that the boost output component looks quite cluttered due to the nested components. 

If there is only going to be one Boost output, then the view plan component can be removed and the distance/power can be included in the boost output component directly rather than as a table.

If there can be many Boost outputs (multiple files), then I think having boost output as a separate card would be useful, implementing the view plan component as similar to the "power plan", "rider" and "bike" accordions in boost configuration

I also noticed there was a `client/src/components/common/boost/test ` file, not sure if that was intended to be pushed.

**Comment by kdav108** [comment]
> I also noticed there was a `client/src/components/common/boost/test ` file, not sure if that was intended to be pushed.

Yep that was a mistake, probably a consequence of using `git add .` 😂 Fixed now.

> The main thing I can see is that the boost output component looks quite cluttered due to the nested components.
>
>If there is only going to be one Boost output, then the view plan component can be removed and the distance/power can be included in the boost output component directly rather than as a table.

Fair point. We would only have one Boost output, but the reason why that 'View plan' drop down was added was to avoid cluttering the UI. If we add the table directly, I feel it's not as pleasant as a user to select the configs from "power plan", "rider", etc. and then having to scroll back up to hit the `Generate` button, because the Boost Output card separates them quite a bit.

I'm open to any suggestions of improving the UI to ensure that both the UI and the code aren't too clustered.

File: client/src/components/common/boost/BoostResults.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  return (
+    <Card className="mt-4">
+      <Card.Body>
+        <Card.Title className="small">
+          <b>BOOST Output:</b>
+        </Card.Title>
+        <div className="small pb-3">
+          Maximum Speed
+          <span className="float-right pr-4">
+            {results ? `${results.maxSpeed} km/h` : 'N/A'}
```
Comment (hallgchris) [inline client/src/components/common/boost/BoostResults.tsx]: Can you format this to two decimal places? Similarly for the distance and power, but these can probably be rounded to integers.

File: client/src/components/common/boost/BoostResults.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * @returns JSX Element
+ */
+export default function BoostResult(props: BoostResultProps) {
+  const { results } = props;
+
+  return (
+    <Card className="mt-4">
+      <Card.Body>
+        <Card.Title className="small">
+          <b>BOOST Output:</b>
```
Comment (hallgchris) [inline client/src/components/common/boost/BoostResults.tsx]: For consistency, maybe drop the `:`

File: client/src/components/common/boost/BoostResults.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  results: BoostResultT | null;
+}
+
+/**
+ * Displays the Maximum achievable speed and the power plan (the distance and the associated recommended power)
+ *
+ * @param props BoostResultProps
+ * @returns JSX Element
+ */
+export default function BoostResult(props: BoostResultProps) {
```
Comment (hallgchris) [inline client/src/components/common/boost/BoostResults.tsx]: Is this component called `BoostResult` or `BoostResults`? There's a discrepancy between the definition here, the filenames, and the names it is imported as.

File: client/src/components/common/boost/BoostConfigurator.tsx [hunk 4]
```diff
@@ -191,6 +201,7 @@ export default function BoostConfigurator({
             >
               Upload All Configs
             </Button>
+            <BoostResult results={boostResult} />
```
Comment (hallgchris) [inline client/src/components/common/boost/BoostConfigurator.tsx]: How would you feel about hiding this completely if `boostResult` is `null`? Alternatively, we could provide a short message saying there is no current plan generated.

**Review by hallgchris — CHANGES_REQUESTED** [review]
A few minor suggestions here!

Regarding the cluttered UI, I agree and have a few suggestions, not that you should necessarily take them.
- You could split the Configuration card into two cards side-by-side, or simply put two columns in the card. Right now there's a lot of unused horizontal space and I don't think it'd make the page any longer by doing this.
- Alternatively, we could, make the entire Boost Output section an accordion.
- Perhaps the Boost Output section doesn't need it's own card-within-a-card? Simply removing one of the nested outlines might clean things up a fair bit.

**Comment by kdav108** [comment]
We can chat about the UI changes in our meeting today, I think it'll much be easier and efficient that way.

**Review by Blake-Haydon — APPROVED** [review]
The code looks good to me. Cant seem to test the change however as I get an error when trying to connect to the server.

```
[HPM] Error occurred while trying to proxy request /sockjs-node from localhost:3000 to ws://localhost:5000 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors) 
```

**Comment by kdav108** [comment]
> The code looks good to me. Cant seem to test the change however as I get an error when trying to connect to the server.
> 
> ```
> [HPM] Error occurred while trying to proxy request /sockjs-node from localhost:3000 to ws://localhost:5000 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors) 
> ```

Did you start the server on a seperate terminal?

**Comment by Blake-Haydon** [comment]
> > The code looks good to me. Cant seem to test the change however as I get an error when trying to connect to the server.
> > ```
> > [HPM] Error occurred while trying to proxy request /sockjs-node from localhost:3000 to ws://localhost:5000 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors) 
> > ```
> 
> Did you start the server on a seperate terminal?

Yeah server running as normal. Might try and re download the repo.

**Comment by Blake-Haydon** [comment]
Works fine now after reinstalling everything. I think the issue occurs when the sever is already running and the client is loading. All good now @kdav108. 

Just testing it then and it works really nicely! Good job!

File: client/src/components/common/boost/BoostResults.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @param props BoostResultProps
+ * @returns JSX Element
+ */
+export default function BoostResults(props: BoostResultsProps) {
+  const { results } = props;
+  const speedPrecision = 2;
+
+  if (!results) {
+    return <div>No Power Plan Profile Generated Yet...</div>;
```
Comment (hallgchris) [inline client/src/components/common/boost/BoostResults.tsx]: This looks a bit odd with each word capitalised, consider using sentence case

File: client/src/components/common/boost/BoostConfigurator.tsx [hunk 6]
```diff
@@ -247,6 +258,24 @@ export default function BoostConfigurator({
               </Card>
             ))}
           </Accordion>
+          <Accordion className="small p-0 my-4">
+            <Card>
+              <Accordion.Toggle
```
Comment (hallgchris) [inline client/src/components/common/boost/BoostConfigurator.tsx]: I reckon it'd be good to disable this toggle (hopefully that's something built into bootstrap) when there's no results - cleaner than having to open it up to reveal that there's nothing inside

**Review by hallgchris — CHANGES_REQUESTED** [review]
A few minor suggestions.

`socket.js` - Can remove lines 111-113, 266-272

Also, although this can be in another PR, we really need to make dashboard retain retained MQTT messages for BOOST. Right now I need to restart BOOST in order for configs to show up after a page refresh, and likewise generated plans disappear after refreshing. If it's easy it could be added here.

File: client/src/components/common/boost/BoostConfigurator.tsx (hunk 6 above)
Comment (kdav108) [reply client/src/components/common/boost/BoostConfigurator.tsx]: My only concern was that if it's not visible initially, then once generated people may not know to scroll down to view this toggle. If it's there from the start they know that the result are stored inside this button. How about disabling the button so it's not clickable when there are no results?

Comment (hallgchris) [reply client/src/components/common/boost/BoostConfigurator.tsx]: Sorry I wasn't clear - that's exactly what I was thinking

Comment (kdav108) [reply client/src/components/common/boost/BoostConfigurator.tsx]: I found out that disabling is not a feature in Accordion, we would have to workaround it with our own code built on top. I think that would increase the complexity for not much benefit, based off some Stackoverflow posts. So for now I've just made it appear only when results are there. Let me know if that's okay.

Comment (hallgchris) [reply client/src/components/common/boost/BoostConfigurator.tsx]: Should be fine!

**Review by hallgchris — APPROVED** [review]
Just remove the lines mentioned in my last review and we're good!

**Pull request merged**